File: src/main/java/com/dgsystems/kanban/entities/CardList.java
package com.dgsystems.kanban.entities;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public record CardList(UUID id, String title, List<Card> cards) {
    public CardList add(Card card) {
        List<Card> updated = new ArrayList<>(cards);
        updated.add(card);
        return new CardList(id, title, updated);
    }

    public CardList remove(Card card) {
        List<Card> updated = new ArrayList<>(cards);
        updated.remove(card);
        return new CardList(id, title, updated);
    }
}

File: backend_elixir/kanban_engine/lib/kanban_engine/CardList.ex
defmodule KanbanEngine.CardList do
  alias KanbanEngine.CardList

  defstruct [:id, :title, :cards]

  def new(id, title) do
    %CardList{id: id, title: title, cards: %{}}
  end

  def add(cardlist, card) do
    %{cardlist | cards: Map.put(cardlist.cards, card.id, card)}
  end

  def remove(cardlist, card) do
    %{cardlist | cards: Map.delete(cardlist.cards, card.id)}
  end
end
